src/ledger/LedgerManagerImpl.cpp

```diff
@@ -1759,6 +1762,399 @@ LedgerManagerImpl::prefetchTransactionData(AbstractLedgerTxnParent& ltx,
+                    it->second = {updatedLe, true};
+                }
+            }
+
+            for (auto const& key : res.getRestoredKeys().hotArchive)
+            {
+                auto [_, inserted] = threadRestoredKeys.hotArchive.emplace(key);
+                releaseAssert(inserted);
+            }
+            for (auto const& key : res.getRestoredKeys().liveBucketList)
+            {
+                auto [_, inserted] =
+                    threadRestoredKeys.liveBucketList.emplace(key);
+                releaseAssert(inserted);
+            }
+        }
+        else
+        {
+            releaseAssert(!txBundle.getResPayload()->isSuccess());
+        }
+    }
+
+    for (auto const& kvp : readOnlyTtlExtensions)
+    {
+        auto const& lk = kvp.first;
+        auto const& ttlExtension = kvp.second;
+
+        auto it = entryMap->find(lk);
+        // The entry in entryMap should always exist. If the entry was deleted,
+        // then we would've erased the TTL key from readOnlyTtlExtensions.
+        releaseAssert(it != entryMap->end() && it->second.mLedgerEntry);
+        if (it->second.mLedgerEntry->data.ttl().liveUntilLedgerSeq <
+            ttlExtension)
+        {
+            it->second.mLedgerEntry->data.ttl().liveUntilLedgerSeq =
+                ttlExtension;
+            it->second.isDirty = true;
+        }
+    }
+
+    return {threadRestoredKeys, std::move(entryMap)};
+}
+
+ParallelLedgerInfo
+getParallelLedgerInfo(AppConnector& app, AbstractLedgerTxn& ltx)
+{
+    auto const& lh = ltx.loadHeader().current();
+    return {lh.ledgerVersion, lh.ledgerSeq, lh.baseReserve,
+            lh.scpValue.closeTime, app.getNetworkID()};
+}
+
+void
+LedgerManagerImpl::applySorobanStage(AppConnector& app, AbstractLedgerTxn& ltx,
+                                     ApplyStage const& stage,
+                                     Hash const& sorobanBasePrngSeed)
+{
+    for (auto const& txBundle : stage)
+    {
+        // This can mark a tx as failed due to validation.
+        // Make sure results are checked.
+        txBundle.getTx()->preParallelApply(
+            app, ltx, txBundle.getEffects().getMeta(), txBundle.getResPayload(),
+            txBundle.getEffects().getTxEventManager());
+    }
+
+    auto const& config = app.getConfig();
+    auto const& sorobanConfig = getSorobanNetworkConfigForApply();
+
+    std::vector<
+        std::future<std::pair<RestoredKeys, std::unique_ptr<ThreadEntryMap>>>>
+        threadFutures;
+
+    auto ledgerInfo = getParallelLedgerInfo(app, ltx);
+    for (size_t i = 0; i < stage.numClusters(); ++i)
+    {
+        auto const& cluster = stage.getCluster(i);
+
+        auto entryMapPtr = collectEntries(app, ltx, cluster);
+
+        threadFutures.emplace_back(std::async(
+            std::launch::async, &LedgerManagerImpl::applyThread, this,
+            std::ref(app), std::move(entryMapPtr), std::ref(cluster), config,
```

Comment:
std::cref just in case (or ideally, a smart pointer to const)

Reply:
Fair point, although I think this isn't _just_ about lifetime. I'm hoping to have a very clear data transfer contract between threads. For example, with const ref, the outer function may mistakenly mutate the values while some thread is referencing them.  The safest pattern here would be create "setup" data in the outer function -> pass ownership to a thread, let the thread mutate encapsulated data that no one else has access to -> pass the result/ownership back to main.

Reply:
Updated